File: Java/src/Array_Hashing/EncodeAndDecodeStrings.java
import java.util.ArrayList;
import java.util.List;

public class EncodeAndDecodeStrings {
    public String encode(List<String> strs) {
        String result = "";

        for (String word : strs) {
            result += word.length() + "#" + word;
        }

        return result;
    }

    public List<String> decode(String encodedString) {
        List<String> result = new ArrayList<>();
        int index = 0;

        while (index < encodedString.length()) {
            int nextHashtag = encodedString.indexOf('#', index);
            String wordLengthText = encodedString.substring(index, nextHashtag);

            int wordLength = Integer.parseInt(wordLengthText);
            int wordStart = nextHashtag + 1;
            String word = encodedString.substring(
                wordStart,
                wordStart + wordLength
            );

            result.add(word);
            index = wordStart + wordLength;
        }

        return result;
    }
}

File: TypeScript/ArraysAndHashing/EncodeAndDecodeStrings.ts
class Solution {
    encode(strs: string[]): string {
        let result = "";

        for (const word of strs) {
            result += word.length + "#" + word;
        }

        return result;
    }

    decode(encodedString: string): string[] {
        const result: string[] = [];
        let index = 0;

        while (index < encodedString.length) {
            const nextHashtag = encodedString.indexOf("#", index);
            const wordLengthText = encodedString.slice(index, nextHashtag);

            const wordLength = Number(wordLengthText);
            const wordStart = nextHashtag + 1;
            const word = encodedString.slice(
                wordStart,
                wordStart + wordLength
            );

            result.push(word);
            index = wordStart + wordLength;
        }

        return result;
    }
}
